### ISSUE TITLE
Error while cutting complex curve.

### ISSUE THREAD

**Issue by blurfl** [MaslowCNC/GroundControl#300]
I moved this here from Firmware#234 

Using GC with PR#298 and Firmware from PR#233 to cut a file that amounts to five 12MM diameter pocket cuts 3mm deep. I originally saw the error using metric coordinates so I created a new version using inch values. The error occurred with the inch version as well, though at a slightly differewnt line in the file.
Home has been moved 5" in the Y direction. The error seems to have occurred while executing line 54, the cutter headed west, clearly in error.
Here is the section of the file showing that line:
<img width="637" alt="gc view gcode see line 54" src="https://cloud.githubusercontent.com/assets/1851315/26531454/438d6b4e-439e-11e7-86a4-5b7fcd52240c.png">

Here is the control panel showing the interpretation. The last correct line looks like it was line 53, the following line is incorrect:
<img width="745" alt="gc executing gcode ending with line 54" src="https://cloud.githubusercontent.com/assets/1851315/26531457/55ad9ec0-439e-11e7-97ac-8cf7ba0e291c.png">

This is the file in use when the screen shots were made:
[Magnet array 12mm x 6cm x 5-SAE.nc.zip](https://github.com/MaslowCNC/GroundControl/files/1034425/Magnet.array.12mm.x.6cm.x.5-SAE.nc.zip)

**Comment by davidelang** [comment]
there was a bug in GC that caused buffer overflow errors. This was fixed late 
yesterday, check that you have the latest version, and check to see if you have 
any errors show up while you are cutting.

**Comment by blurfl** [comment]
That was PR#298, the version I'm using. First thing I did when the error happened was to check the pertinent file for the change, yes.

**Comment by BarbourSmith** [comment]
It's very interesting that just the Y term disappeared completely. That could be a Ground Control issue, but it could also be a firmware thing. Let's keep the issue open here. It's so nicely documented, it would be a shame to move it.

**Comment by blurfl** [comment]
On different runs I've seen the z term lost, the decimal point and part of the mantissa lost, and one time where only the letter Y was there. Was there a change to a buffer routine, or are we running low on ram?

**Comment by davidelang** [comment]
there have been changes to the buffering in the last couple of weeks. There were 
errors that bar fixed in this week's release, but he may not have got all of 
them.

loosing the tail end of the command does sound like a buffer problem.

**Comment by BarbourSmith** [comment]
I've actually noticed the '\n' (an invisible character that marks the end of the line) getting lost pretty regularly too.

I'll start testing by running the file attached here (Thank you!)

If I can't track it down that way, I'll make a firmware GC pair where we just send a known line back and forth as fast as possible and check that it arrives correctly on both ends

**Comment by davidelang** [comment]
hmm, if we're loosing \n routinely, that makes me think there may be an 
off-by-one error in there somewhere.

**Comment by blurfl** [comment]
I've started looking into Firmware PR#229, the returnError() routine. I wonder if that's too much function calling and printing going on. I've figured out that I can't just comment the print statements out - because the ringBuffer.spaceAvailable() value gets used in serialPortThread.py - but I could shorten things by replacing the previous two function calls and associated print statements with "0,0,". This still has the problem, though, so not the solution.

**Comment by BarbourSmith** [comment]
I've got some bad news which is that I'm not seeing the same behavior. I've run the file on three different places on the sheet, and in all three it ran to the end without incident. 

This is going to make it harder for us to figure out what is going on. I would bet that the issue is in some way dependent on the computer itself (the only thing different between our setups). That doesn't mean it's not a problem with our software, just that it's a problem which only pops up under certain circumstances. 

It makes sense to me that we would see this issue now and with this file. The file has a large number of small lines of gcode which are executed very quickly. 

When you have a chance @blurfl give the branch [fix for 300](https://github.com/MaslowCNC/GroundControl/tree/fix-for-300) a try. I've added some print statements to keep an eye on the health of the computers internal serial buffers for both transmitting and receiving.

The **in** number indicates how many bytes of data are waiting to be processed in the incoming buffer and the **out** number indicates how many bytes of data are waiting to be processed on the out going buffer. On my system the out number stays at 0 pretty much all the time because the arduino is pretty good about reading new data right away, while the in hovers under 100.

![image](https://cloud.githubusercontent.com/assets/9359447/26739912/73bfaab4-4788-11e7-87ab-4abeac04bd49.png)

What do you see on your end? Is it similar? Do those numbers change at all when you see the error crop up?

**Comment by blurfl** [comment]
Good news, actually - it's another data point! I'll try the branch you ask, and if I need to, I'll fire up the RPi and see how it goes on that platform. If need be, I'll drag the Win10 VM out to the shop. 
It'll be later tonight before I can get to it, but I'll let you know what I see. Thanks for giving it your attention!

**Comment by blurfl** [comment]
I took a moment to install the 'fix-for-309' version of Firmware and run the 'fix-for-300' version of GC with the .nc file. I put the terminal window next to the GC window and stopped as quickly as I could after the error. I've drawn an arrow between the two to show which line I think is in question. From your description of the serial buffer counters, they look normal, yes?

![error using gc-fix-for-300](https://cloud.githubusercontent.com/assets/1851315/26749527/ca72edc4-47c1-11e7-837b-8d3e51442711.jpg)

**Comment by blurfl** [comment]
I really do think this issue is related to the buffering code added to the Firmware in release 72. 
Trying to test this, I've gone back to earlier release pairs. So far I've tried every GC/Firmware release pair since F69/GC71 and my test file works correctly in all except F72/GC74. To test whether the issue was GC74 or the Firmware72, I checked with F71 and GC74 and there was no error. To check for a baud rate problem in F72/GC74 I've tried 19200 baud and themerror occurred. I've done a successful run at 115200 baud using GC74 with F71, but the error crops up if I then install F72. (I've seen a suggestion on a different CNC site that low baud rates actually slow down an Arduino Uno.)
By this I think I've narrowed the issue down to something in F72, and something other than the baud rate. 
I was going to try F72/GC74 on a RPi but I find that my RPi posts python errors with GC74. I haven't taken time to step back to see whether previous releases work on my RPi setup - its been a couple months since I last used it.

**Comment by davidelang** [comment]
This is exactly the sort of thing that git bisect was designed for 
https://git-scm.com/docs/git-bisect

**Comment by blurfl** [comment]
That's an interesting tool, thanks. Much to learn, CNC is just a part of it 🙂

**Comment by blurfl** [comment]
I modified the Firmware to echo each character received out the Serial3 port, which comes out AUX4 on the motor board. Capturing this stream, I see that the error is coming from GC and not the Firmware. The stream seen on Serial3 is the same as that executed and echoed back to GC by the Firmware.
 To check whether the fault is hardware or software, I tested with Linux running on a different laptop and found the same thing, characters missing from the serial stream. I changed cables and saw no difference so the cables are nto the cause. I ran the .nc file using UGC on both platforms and the cut was correct in both cases. When I switch releases back to F71GC73, the cut is correct on both platforms. 
 If I were a better pythonista, I would have GC.serialPortThread.py log the characters to a file as it writes them to the serial port. There is a buffer in GC.serialPortThread.py, perhaps this one is the culprit?

**Comment by blurfl** [comment]
I wonder whether we are swamping the pyserial out buffer. Serial.outWaiting() is available to check for buffer space, but that's over my head... A higher baud rate might drain the buffer faster, I'll try that.

**Comment by davidelang** [comment]
well, right now we seem to have a problem where Ground Control is not outputting 
what it thinks it is. We need to get logs to show what it thinks it's sending 
out vs what the thing on the other end of the serial port says that it's 
getting.
k

**Comment by blurfl** [comment]
OK, I've found the output of the print statement in  GC.serialPortThread.py _write() during an error:

   ding: G1 X146.315685279 Y60.0479695431
   ding: G1 X146.275076142 Y60.1418781726
   ding: G1 X146.216700508 Y60.2230964467
   ding: G1 X146.143096447 Y60.2967005076
   ding: G1 X146.061878173 Y60.3550761421
   ding: G1 X145.967969543 Y60.3956852792

And on the motor board, the serial echo sends:

G1 X146.315685279 Y60.0479695431
G1 X146.275076142 Y60.1418781726
G1 X14G1 X146.143096447 Y60.2967005076
G1 X146.061878173 Y60.3550761421
G1 X145.967969543 Y60.3956852792

The third line seems to have been correct when GC passed it to serialInstance.write but truncated when it arrived at the other end of the wire.

Note: I repeated this test at 230400 baud to see whether draining the buffer faster helped. It did not (though otherwise, the communication seemed to keep up at that rate :) ).

**Comment by davidelang** [comment]
can you tweak the software to display how much buffer space it thinks there is 
before each write (and ideally the position in the buffer)?

how repeatable is this error?
k

**Comment by blurfl** [comment]
To be more complete, I'll include logs from the last Z movement, because during this part of the cut the GC buffer check triggered:

---------------------------------------------

Lines sent by GC:

   ding: G0 Z5
   ding: G0 X145.324314721 Y60.2241116751
   ding: G1 Z-1.8 F800
   ding: G1 X145.264923858 Y60.1418781726 F1500
   ding: G1 X145.224314721 Y60.0479695431
Run Paused
Unable to check buffer
[PosError:0.302,211]
Run Resumed
   ding: G1 X145.196395939 Y59.9515228426
   ding: G1 X145.188781726 Y59.85
   ding: G1 X145.196395939 Y59.7484771574
   ding: G1 X145.224314721 Y59.6520304569
   ding: G1 X145.264923858 Y59.5581218274
   ding: G1 X145.323299492 Y59.4769035533
   ding: G1 X145.396903553 Y59.4032994924
   ding: G1 X145.478121827 Y59.3449238579
   ding: G1 X145.572030457 Y59.3043147208
   ding: G1 X145.668477157 Y59.2763959391
   ding: G1 X145.77 Y59.2687817259
   ding: G1 X145.871522843 Y59.2763959391
   ding: G1 X145.967969543 Y59.3043147208
   ding: G1 X146.061878173 Y59.3449238579
   ding: G1 X146.143096447 Y59.4032994924
   ding: G1 X146.216700508 Y59.4769035533
   ding: G1 X146.275076142 Y59.5581218274
   ding: G1 X146.315685279 Y59.6520304569
   ding: G1 X146.343604061 Y59.7484771574
   ding: G1 X146.351218274 Y59.85
   ding: G1 X146.343604061 Y59.9515228426
   ding: G1 X146.315685279 Y60.0479695431
   ding: G1 X146.275076142 Y60.1418781726
   ding: G1 X146.216700508 Y60.2230964467
   ding: G1 X146.143096447 Y60.2967005076
   ding: G1 X146.061878173 Y60.3550761421
   ding: G1 X145.967969543 Y60.3956852792
   ding: G1 X145.871522843 Y60.4236040609
   ding: G1 X145.77 Y60.4312182741
Gode Stopped
Sending: 

 ---------------------------------------------
 

 Lines received by motor board serial echo:
  
  
G0 Z5
G0 X145.324314721 Y60.2241116751
G1 Z-1.8 F800
G1 X145.264923858 Y60.1418781726 F1500
G1G1 X145.196395939 Y59.9515228426
G1 X145.188781726 Y59.85
G1 X145.196395939 Y59.7484771574
G1 X145.224314721 Y59.6520304569
G1 X145.264923858 Y59.5581218274
G1 X145.323299492 Y59.4769035533
G1 X145.396903553 Y59.4032994924
G1 X145.478121827 Y59.3449238579
G1 X145.572030457 Y59.3043147208
G1 X145.668477157 Y59.2763959391
G1 X145.77 Y59.2687817259
G1 X145.871522843 Y59.2763959391
G1 X145.967969543 Y59.3043147208
G1 X146.061878173 Y59.3449238579
G1 X146.143096447 Y59.4032994924
G1 X146.216700508 Y59.4769035533
G1 X146.275076142 Y59.5581218274
G1 X146.315685279 Y59.6520304569
G1 X146.343604061 Y59.7484771574
G1 X146.351218274 Y59.85
G1 X146.343604061 Y59.9515228426
G1 X146.315685279 Y60.0479695431
G1 X146.275076142 Y60.1418781726
G1 X14G1 X146.143096447 Y60.2967005076
G1 X146.061878173 Y60.3550761421
G1 X145.967969543 Y60.3956852792
G1 X145.871522843 Y60.4236040609
G1 X145.77 Y60.4312182741
!
   
 ---------------------------------------------

**Comment by blurfl** [comment]
Thanks for remindsing me, @BarbourSmith has a [branch ](https://github.com/MaslowCNC/GroundControl/tree/fix-for-300) that does that - I'll capture that as well.

**Comment by davidelang** [comment]
Ok, the good news is that this erro correlates to the "unable to check buffer" 
error (which makes sense), so why did that error trigger?

**Comment by davidelang** [comment]
hmm, why are we inventing a new protocol rather than just using xon/xoff 
signaling?

we shouldn't have to guess how much space is left in the buffer, we should just 
turn on xon/xoff for the serial connection in python and then implement it on 
the arduino side (and I think it's stupid for the arduino hardware to not 
support hardware flow control :-)

**Comment by blurfl** [comment]
Using the 'fix-for-300' branch, these logs:


Lines logged by GC:

   ------------------------------------------


Connected on port /dev/tty.usbmodem14441
in: 80	- out: 	0  : B03 A2438.4 C1219.2 Q2995.65 E369.29 F310.2 R139 H79 I1 J8148.0 K10 M6.35 N3.17 P7560.0   
in: 121	- out: 	0  : B05   
in: 176	- out: 	0  : G21   
in: 26	- out: 	0  :    
in: 36	- out: 	0  :   
    45	- out: 	0  : G21 G90 G40
in: 28	- out: 	0  :   
in: 17	- out: 	0  :   
    48	- out: 	0  : G0 Z5
    47	- out: 	0  : T0 M6
    12	- out: 	0  : G17
    205	- out: 	0  : M3
    77	- out: 	0  : G0 X151.518984772 Y104.223096447
    51	- out: 	0  : G1 Z-1.8 F800
    20	- out: 	0  : G1 X151.422030457 Y104.195685279 F1500
    0	- out: 	0  : G1 X151.328121827 Y104.155076142
    20	- out: 	0  : G1 X151.246903553 Y104.096700508
    66	- out: 	0  : G1 X151.173299492 Y104.023096447
    129	- out: 	0  : G1 X151.114923858 Y103.941878173
    65	- out: 	0  : G1 X151.074314721 Y103.847969543
    95	- out: 	0  : G1 X151.046395939 Y103.751522843
    26	- out: 	0  : G1 X151.038781726 Y103.65
    7	- out: 	0  : G1 X151.046395939 Y103.548477157
    174	- out: 	0  : G1 X151.074314721 Y103.452030457
    167	- out: 	0  : G1 X151.114923858 Y103.358121827
    102	- out: 	0  : G1 X151.173299492 Y103.276903553
    102	- out: 	0  : G1 X151.246903553 Y103.203299492
    87	- out: 	0  : G1 X151.328121827 Y103.144923858
    18	- out: 	0  : G1 X151.422030457 Y103.104314721
    6	- out: 	0  : G1 X151.518477157 Y103.076395939 <---EOL lost
    38	- out: 	0  : G1 X151.62 Y103.068781726
    73	- out: 	0  : G1 X151.721522843 Y103.076395939
    129	- out: 	0  : G1 X151.817969543 Y103.104314721
    112	- out: 	0  : G1 X151.911878173 Y103.144923858
    74	- out: 	0  : G1 X151.993096447 Y103.203299492
    5	- out: 	0  : G1 X152.066700508 Y103.276903553
    19	- out: 	0  : G1 X152.125076142 Y103.358121827
    6	- out: 	0  : G1 X152.165685279 Y103.452030457
    66	- out: 	0  : G1 X152.193604061 Y103.548477157
    21	- out: 	0  : G1 X152.201218274 Y103.65
    31	- out: 	0  : G1 X152.193604061 Y103.751522843
    127	- out: 	0  : G1 X152.165685279 Y103.847969543
    174	- out: 	0  : G1 X152.125076142 Y103.941878173
    181	- out: 	0  : G1 X152.066700508 Y104.023096447
    181	- out: 	0  : G1 X151.993096447 Y104.096700508
    194	- out: 	0  : G1 X151.911878173 Y104.155076142
    125	- out: 	0  : G1 X151.817969543 Y104.195685279
    101	- out: 	0  : G1 X151.721522843 Y104.223604061
    75	- out: 	0  : G1 X151.62 Y104.231218274
    6	- out: 	0  : G1 X151.518477157 Y104.223604061
    19	- out: 	0  : G1 X151.187005076 Y106.106345178 F1500 <--- chars lost between Y1 & EOL
    6	- out: 	0  : G1 X150.767208122 Y105.992639594 F1500
    5	- out: 	0  : G1 X150.373807107 Y105.809898477
    22	- out: 	0  : G1 X150.015939086 Y105.561167513
Gode Stopped
in: 26	- out: 	0  : !  


   ------------------------------------------
 

Lines logged by motor board:

G21 G90 G40
  
  
G0 Z5
T0 M6
G17
M3
G0 X151.518984772 Y104.223096447
G1 Z-1.8 F800
G1 X151.422030457 Y104.195685279 F1500
G1 X151.328121827 Y104.155076142
G1 X151.246903553 Y104.096700508
G1 X151.173299492 Y104.023096447
G1 X151.114923858 Y103.941878173
G1 X151.074314721 Y103.847969543
G1 X151.046395939 Y103.751522843
G1 X151.038781726 Y103.65
G1 X151.046395939 Y103.548477157
G1 X151.074314721 Y103.452030457
G1 X151.114923858 Y103.358121827
G1 X151.173299492 Y103.2769035599492
G1 X151.328121827 Y103.144923858
G1 X151.422030457 Y103.104314721
G1 X151.518477157 Y103.076G1 X151.62 Y103.068781726  <------ error here
G1 X151.721522843 Y103.076395939
G1 X151.817969543 Y103.104314721
G1 X151.911878173 Y103.144923858
G1 X151.993096447 Y103.203299492
G1 X152.066700508 Y103.276903553
G1 X152.125076142 Y103.358121827
G1 X152.165685279 Y103.452030457
G1 X152.193604061 Y103.548477157
G1 X152.201218274 Y103.65
G1 X152.193604061 Y103.751522843
G1 X152.165685279 Y103.847969543
G1 X152.125076142 Y103.941878173
G1 X152.066700508 Y104.023096447
G1 X151.993096447 Y104.096700508
G1 X151.911878173 Y104.155076142
G1 X151.817969543 Y104.195685279
G1 X151.721522843 Y104.223604061
G1 X151.62 Y104.231218274
G1 X151.518477157 Y104.223604061
G1 X151.187005076 Y1                   <-------- error here 
G1 X150.767208122 Y105.992639594 F1500
G1 X150.373807107 Y105.809898477
G1 X150.015939086 Y105.561167513
!

**Comment by blurfl** [comment]
Flow control could help, but since it's not part of the library, it's still "roll-your-own".

**Comment by davidelang** [comment]
digging into the xon/xoff idea, it looks like the arduino boards are broken enough to not support it https://github.com/grbl/grbl/issues/50

It looks like we need to increase the safety factor and try not to send data up to the full capacity of the receive buffer. the link above talks about sending up to 64 characters in a 'packet' via the usb interface, so what if we were to bump up the size of the receive buffer by 64 characters in the firmware, but leave it set as-is on the GC side? this seems wasteful, but may be what we need to do to avoid this sort of grief.

**Comment by blurfl** [comment]
Looking into how they handled comm. in grbl is a good idea.

**Comment by blurfl** [comment]
FWIW, I just tried stepping down to 9600baud. No joy.

**Comment by davidelang** [comment]
make sense. I don't thin this is a timing problem, I think it's an artifact of 
using USB to send data to simulate a serial port. Serial ports handle characters 
one at a time (with handshaking happening per-character), the USB protocol has a 
lot of overhead per packet, so it sends a bunch of characters at a time.

Trying to have the protocol keep the two sides in sync so that they both know 
the size of the buffer and never send more data than there is room to hold is 
fragile, that's why I suggest making the receive buffer noticably larger so that 
if too much data is sent, it doesn't get lost.
k

**Comment by blurfl** [comment]
Just read about upping the Rx buffer on the Arduino [here](https://internetofhomethings.com/homethings/?p=927). I set both Rx and Tx buffers up to 256, no joy.

**Comment by blurfl** [comment]
I think we need to throttle the pyserial end, before the USBserial driver's buffer.

**Comment by blurfl** [comment]
@BarbourSmith, the missing EOLs that you see are an example of the issue.

**Comment by BarbourSmith** [comment]
The good news is we know we (I) introduced this problem a in the last few weeks of big serial changes so we know that the problem can be solved.

I think the first step is to find out exactly which line change caused the issue to first appear. Then we can deduce from that how to fix it.

Which was the last version which worked consistently?

**Comment by blurfl** [comment]
I went back to Firmware 70/GC72, and now that I know what to look for, a truncated line like the one that starts G1 X16.76 here:  - 

<img width="335" alt="screen shot 2017-06-05 at 12 26 20 pm" src="https://cloud.githubusercontent.com/assets/1851315/26799386/45a86436-49ea-11e7-8a09-47b314f5af4b.png">

This line came in from the serial port as:
G1 X16.7639593909 Y108.34010G1 X16.0 Y108.406091371
Looking at the hex dump of that serial input stream, there are no characters between 'Y108.34010' and 'G1 X16.0'.

Nevertheless, now that I know how to spot it on the GC control panel, I can see evidence of missing characters as far back as  Firmware 70/GC72.

**Comment by BarbourSmith** [comment]
That will be a big help in finding what's going on.

If you go back further the problem disappears?

**Comment by blurfl** [comment]
Odd thing, that. With Firmware0.69 loaded running GC71, the motor test works, the z axis works, but the right and left motors don't move. Same situation on the Linux box. A puzzler...

**Comment by BarbourSmith** [comment]
Hmm, that is a puzzler. 

Could it have something to do with the fact that we've gotten rid of the motor calibration process? Maybe that needs to be run again.

**Comment by davidelang** [comment]
The motor calibration process was creating a static power table, the PID loop 
eliminates the need for that. If running the motor calibration process 'fixes' 
anything, it's only because of some unplanned side-effect.

**Comment by blurfl** [comment]
Thanks for the reminder. Motor calibration seems to be the answer. I'll run the cut a little later after the calibration has finished.

**Comment by blurfl** [comment]
The pair Firmware69/GC71 works without the error. Each of the release pairs after that exhibit the error. Here is a file that causes the error more oftern, probably because of the length of the mantissa for each value :) .
[Magnet Circles Sketchup2.nc.zip](https://github.com/MaslowCNC/GroundControl/files/1053432/Magnet.Circles.Sketchup2.nc.zip)

**Comment by davidelang** [comment]
can you try git bisect between that pair and the next version?

commit 3b4e400b62b3073cd34347e8cc9ba38b2ea28d57 is "push lines of gcode until 
buffer is full", which sounds like it's probably the issue, but there were 
several updates to the serial communication in this release.

I suspect this is going to be a case of "yep it broke something" and we will be 
back to troubleshooting exactly why.

This is when the ringbuffer was introduced, allowing multiple commands to be 
sent and buffered, etc.

**Comment by blurfl** [comment]
It sound like you're the most experienced with git bisect. Can you do this? Would probably be speedier, yes?

 I'm only a puppy here; I've gotten this far:
git bisect start
git bisect good 61391096330224c7a84fd1e17eac4e6e76c2906c // the SHA of PR bump to version71 
git bisect bad cb0da072eb00a37b7714cf035bbb1ca737bc0f8e // the SHA of PR bump to version72

which tells me:
Bisecting: 8 revisions left to test after this (roughly 3 steps)
[59d09f723dc845a4dbf69d3206eed88a8f03565e] experimenting with points

**Comment by davidelang** [comment]
unfortunantly, I have no way to test the result :-(


On Mon, 5 Jun 2017, Scott Smith wrote:

> It sound like you're the most experienced with git bisect. Can you do this? Would probably be speedier, yes?
>
> I'm only a puppy here; I've gotten this far:
> git bisect start
> git bisect good 61391096330224c7a84fd1e17eac4e6e76c2906c // the SHA of PR bump to version71
> git bisect bad cb0da072eb00a37b7714cf035bbb1ca737bc0f8e // the SHA of PR bump to version72
>
> which tells me:
> Bisecting: 8 revisions left to test after this (roughly 3 steps)
> [59d09f723dc845a4dbf69d3206eed88a8f03565e] experimenting with points

so what you need to do now is to run this version and then tell git bisect if it 
works or not, at that point git will checkout a different version.

with 8 commits, you could test them one at a time, git bisect will probably find 
the cutoff in 3 steps (the savings are higher the more commits there are)

**Comment by blurfl** [comment]
Shoot me .zipped files to run?

**Comment by davidelang** [comment]
I had a thought, we are able to use a grbl gcode sender instead of ground 
control to send g-code to the firmware. If you try using that do you have the 
same problem?

If not, that would say it's a GroundControl problem.
If the problem is still there, that would say it's a firmware problem.

This should let us know which side of things we should be looking at.

**Comment by blurfl** [comment]
I ran the .nc file using Universal Gcode Sender on both OS X and Ubuntu17.4 and the cut was correct in both cases. 
I have not yet wired up a different USBserial adapter to take the Arduino drivers out of the equation.

**Comment by davidelang** [comment]
Ok, that makes it look like it's the GroundControl side sending data when there 
isn't enough room in the buffer.

I will say that _checkBufferSize looks odd to me in serialPortThread.py

it has a couple paths that will set self.machineIsReadyForData to true, but 
nothing that will set it to false. If I am reading it correctly (minutes after I 
wake up, so a bit groggy still), if there isn't room in the buffer, it will 
throw the exception (unable to to check buffer), there is no path for when it 
can check the buffer, but there just isn't room right now

It seems to me that if we are doing the queue management on the GC side, we 
should be doing something along the lines of keeping track of the length of each 
message that we have sent, and when we get an 'ok\n' (but we look for 'ok\r\n') 
we should decrease the amount of buffer used by the length of the oldest message 
and check if we can send a new message, repeating every time we get an ok 
message.

but as I say, I'm not fully awake yet, so I could be missing something here.

**Comment by blurfl** [comment]
valz[2][0:-3] is ringbuffer space available as sent from the Firmware. When I patched serialPortThread.py to print this every time it checks to send, I see it most usually at 256, occasionally lower.

**Comment by davidelang** [comment]
right, but that routine is only called when we get a [ back from the firmware, 
the ok is basically ignored, but that is the only time when there should be the 
possibility of more room being available

also remember that we have the error 'unable to check buffer' when we loose 
data, which means that those tests are not passing, but we are still sending 
data.

**Comment by blurfl** [comment]
The value is sent by the firmware in CNC_functions.h each time a gcode line is parsed.

**Comment by davidelang** [comment]
it doesn't matter when it's set in the firmware, what matters is when it's read 
in GC.

according to the thread about grbl, it tracks when a command is completed by 
looking for the ok, not by getting any buffer size message from the firmware.

trying to get buffer size messages from the firmware is going to be unreliable 
because there can be several messages in flight from the firmware to GC, so you 
may be acting on old data if you just decide based on the last thing you read 
from the serial stream.

you have to keep track of the messages you have sent vs the messages that have 
been acknowledged as being completed. otherwise you don't know if the message 
you are reacting to is the most recent one sent, or one from a few exchanges 
back.

**Comment by BarbourSmith** [comment]
Great deduction. This sounds right to me. I didn't like the GRBL approach of counting the number of characters sent because an assumption is made about the length of buffer in the firmware and because you can get out of sync if the gcode sender program misses an "OK". My thinking was that if we're always printing the actual number of characters available things  will stay in sinc. That may have been over thinking things, so let's re-evaluate.

@blurfl is @davidelang correct that you see **'the unable to check buffer'** message when data is lost? Is it every time?

**Comment by davidelang** [comment]
On Tue, 6 Jun 2017, BarbourSmith wrote:

> @blurfl is @davidelang correct that you see **'the unable to check buffer'** message when data is lost? Is it every time?

it happened at least once in the examples posted here.

**Comment by blurfl** [comment]
I haven't been looking for the 'unable...' message, I will run some tests and report. To be sure I'm looking in the correct place - I'll need to run GC from the terminal  command line, and that message will be seen in the terminal command line window, yes?

**Comment by blurfl** [comment]
Ok, I just loaded the current releases and started a run. The issue arose with no sight of 'Unable...' in log.txt or the terminal command line window.

**Comment by blurfl** [comment]
I've snipped the pertinent sections from the serial echo, the terminal window and log.txt with the error line indicated. Note that in log.txt, in the PosError lines the third value is ringbuffer free space.

Serial echo:
G1 X16.0 Y106.494923858
G1 X15.5659898477 Y106.456852792
G1 X15.2347715736 Y108.339593909 F1500
G1 X14.4923857868 Y108.139593909 F1500
G1 X13.7969543147 Y107.817258883
G1 X13.1675126904 Y107.375634518
G1 X12.6243654822 Y106.83248731
G1 X12.1827411168 Y106.20304G1 X11.6598984772 Y104.763959391   <--------------
G1 X11.5939086294 Y104.0
G1 X11.6598984772 Y103.236040609
G1 X11.8604060914 Y102.492385787
G1 X12.1827411168 Y101.796954315


Terminal window:
   ding: G1 X16.0 Y106.494923858
   ding: G1 X15.5659898477 Y106.456852792
   ding: G1 X15.2347715736 Y108.339593909 F1500
   ding: G1 X14.4923857868 Y108.139593909 F1500
   ding: G1 X13.7969543147 Y107.817258883
   ding: G1 X13.1675126904 Y107.375634518
   ding: G1 X12.6243654822 Y106.83248731
   ding: G1 X12.1827411168 Y106.203045685     <----------------
   ding: G1 X11.8604060914 Y105.507614213
   ding: G1 X11.6598984772 Y104.763959391
   ding: G1 X11.5939086294 Y104.0
   ding: G1 X11.6598984772 Y103.236040609
   ding: G1 X11.8604060914 Y102.492385787
   ding: G1 X12.1827411168 Y101.796954315


Log.txt capture:
[PosError:0.51,0.30,176]
G1 X16.0 Y106.494923858
[PosError:0.43,0.38,170]
G1 X15.5659898477 Y106.456852792
<Idle,MPos:16.00,106.49,-1.82,WPos:0.000,0.000,0.000>
[PosError:0.19,0.27,170]
[PosError:0.29,0.53,176]
G1 X15.2347715736 Y108.339593909 F1500
[PosError:0.49,0.07,219]
G1 X14.4923857868 Y108.139593909 F1500
<Idle,MPos:15.23,108.34,-1.82,WPos:0.000,0.000,0.000>
[PosError:0.19,0.02,219]
[PosError:0.44,0.57,256]
G1 X13.7969543147 Y107.817258883
<Idle,MPos:13.80,107.82,-1.82,WPos:0.000,0.000,0.000>
[PosError:0.06,0.11,256]
[PosError:0.01,0.06,255]
G1 X13.1675126904 Y107.375634518
[PosError:0.17,0.74,228]
G1 X12.6243654822 Y106.83248731
[PosError:0.17,0.85,256]
G1 X12.1827411168 Y106.20304                              <----------------
<Idle,MPos:12.62,106.83,-1.82,WPos:0.000,0.000,0.000>
[PosError:0.03,0.57,256]
G1 X11.6598984772 Y104.763959391
[PosError:0.12,1.12,219]
G1 X11.5939086294 Y104.0
<Idle,MPos:11.65,104.62,-1.82,WPos:0.000,0.000,0.000>
[PosError:0.03,0.83,182]
[PosError:0.30,0.79,219]
G1 X11.6598984772 Y103.236040609
[PosError:0.32,0.54,183]
G1 X11.8604060914 Y102.492385787
[PosError:0.37,0.40,183]
G1 X12.1827411168 Y101.796954315

**Comment by BarbourSmith** [comment]
Woah, that's interesting. So the log shows the correct line, but the serial echo shows the line being corrupted. What is the serial echo? Is that a capture of the serial communication?

I'm going to make a branch which adds a final term to each line indicating the correct number of characters which will be checked on the firmware side. I think that could help us track down what's happening and what causes it.

**Comment by blurfl** [comment]
I added a line after Serial.read(c) to print c out Serial3 which is captured by a different USBserial connection. Looking at that serial stream, there are no characters between the digits and the G1 in the middle of the line. The two G1 commands in that line are broken out programmatically by the firmware.

**Comment by BarbourSmith** [comment]
Wow! Great debugging tool!

So we know pretty conclusively at this point that the issue is in Ground Control because we've seen that
1) As soon as the Arduino is reading characters, there are characters missing.
2) Using universal gcode sender fixes the issue.

Right?

**Comment by blurfl** [comment]
Exactly right👌

**Comment by BarbourSmith** [comment]
I'm going to keep working on this check sum idea. I think I'm loosing characters too, but much less than you are. If I can find a reliable test it will help me track down the issue more quickly than having to keep asking you to test things.

**Comment by blurfl** [comment]
On my setup, [this](https://github.com/MaslowCNC/GroundControl/files/1053432/Magnet.Circles.Sketchup2.nc.zip) file causes the error more often than others 😊

**Comment by BarbourSmith** [comment]
I'd run that file before (no issues), and I just ran it again to see if anything had changed. It still cuts without issue, but I am loosing characters from the ends of lines. 

I added a fake check sum where Ground Control adds the length of the line at the end like 'L23' would tell the firmware to expect 23 characters in the line. Interestingly, I only seemed to end up loosing the 'L23' at the end of the line (or part of it).  Loosing one or two characters from the ends of the lines won't usually cause the file to cut noticeably wrong because it's generally just some decimals or the federate or something like that that is lost.

Having the checksum should help me track things down.

If you have a  chance @blurfl will you run that file with [this firmware](https://github.com/MaslowCNC/Firmware/tree/read-check-sum) and [this Ground Control](https://github.com/MaslowCNC/GroundControl/tree/add-check-sum) and let me know what your log looks like?

I'll keep digging on my end.

**Comment by BarbourSmith** [comment]
As a baseline to compare to, I had 20 lines in that file loose at least one character

**Comment by blurfl** [comment]
Losing characters at the end of the line is exactly the issue - if you lose enough characters, Y104.47457543 becomes Y1 and drama ensues😛

**Comment by BarbourSmith** [comment]
Yeah. I ran a cut that went pretty off course one this one:


From the log:

"17-->>Chars lost in serial //

G2 X10.378 Y-6.93

G2 X12.1582 Y-6.8977 I10.1061 J-407.3156  L47 "

That first G2 just cut a big hole in everything.

At least now I can change things and see if they effect the result.

I'm going to start by looking into what changed after GC71 and see if I can track down the root of the issue.

**Comment by BarbourSmith** [comment]
Here's the diff since then for reference https://github.com/MaslowCNC/GroundControl/compare/v0.71...master

**Comment by davidelang** [comment]
even just looking at the diff from v0.71 to v0.72, that is when the ability to 
send lines asynchonously was introduced, so this looks like it has existed since 
the beginning of this feature.

**Comment by davidelang** [comment]
looking at the corresponding firmware diff, that is the cycle when the 
ringbuffer was first introduced.
kk

**Comment by BarbourSmith** [comment]
I think I've gotten the issue to go away by removing the code which checks the ring buffer size and returning to just referencing the 'OK' command. It's slow and there is a noticeable pause between when lines of gcode execute, but it works.

I think the issue has to do with putting too many characters in the send buffer on the python side too quickly. When GC knows the buffer is empty it will dump like 200 characters into the transmit buffer all at once, which I think is causing some of them to get lost in Pyserial

**Comment by BarbourSmith** [comment]
I'm going to try to implement a system which does character counting the same way GRBL does

**Comment by blurfl** [comment]
Is [this](http://www.ftdichip.com/Support/Knowledgebase/index.html?an232beffectbuffsizeandlatency.htm) link about USB serial buffernsize of use/interest?

**Comment by davidelang** [comment]
it explains how there can be multiple messages in flight about the buffer size, 
which would explain how GC could be confused (if the firmware sends 80 bytes 
left, 20 bytes left and GC reads the "80" message and decides to send a 60 byte 
message, data will be lost)

This is why I was saying above that trying to communicate the buffer size isn't 
going to work.

**Comment by BarbourSmith** [comment]
I've re-written how the serial thread works to operate on the same principal as GRBL does where when a line is sent the number of characters are counted, and when the 'OK' message is received for that line the space is freed up.

It works well, but I'm still losing data. If we use the system where we don't send the next line until we see the "OK" from the previous line (eg v0.71), no data is lost. As soon as we start buffering lines in advance we seem to loose some characters. 

Interestingly, on this file I ran we it seems to always be the same line which is lost. 
"20-->>Chars lost in serial //

G3 X1.8217 Y-5.3809 

G3 X1.6884 Y-5.3943 I0.4862 J-5.4951  L43 "


[log.txt](https://github.com/MaslowCNC/GroundControl/files/1056487/log.txt)

**Comment by blurfl** [comment]
Repeatablevis good🙂
If the problem is the USBserial buffer, making sure to take a pause between each 64? characters sent would let the USB packets clear.

**Comment by davidelang** [comment]
I don't buy that, otherwise the grbl universal sender would have the same 
problem

**Comment by BarbourSmith** [comment]
Interestingly when I run @blurfl magnet test file I get no lines reporting missing data down from 20 lines before. There are a couple false alarms in the log, but they don't show the kind of truncation that I was seeing before.

@blurfl When you have a chance, will you give [this Ground Control](https://github.com/MaslowCNC/GroundControl/tree/work-out-what's-going-on) a go and see if it solves the issue for you?

**Comment by blurfl** [comment]
Thinking more, putting in a delay between lines sent might do it. Start with a longer delay and tune it down by testing.

**Comment by davidelang** [comment]
On Tue, 6 Jun 2017, BarbourSmith wrote:

> I've re-written how the serial thread works to operate on the same principal as GRBL does where when a line is sent the number of characters are counted, and when the 'OK' message is received for that line the space is freed up.
>
> It works well, but I'm still losing data. If we use the system where we don't send the next line until we see the "OK" from the previous line (eg v0.71), no data is lost. As soon as we start buffering lines in advance we seem to loose some characters.

double check that you are including the line ending.

as a debugging step, have GC output how much space it thinks are in the buffer, 
and have the firmware output how much space it has in the buffer.

where do the two disagree? is it an error that accumulates until data is lost? 
or is it when you hit a specific amount of data sent (wraparound error of the 
ringbuffer), or something else

minor note, while windows users \r\n as a line ending, everything else 
(including network protocols running on windows, and grbl) use \n, if we are at 
a point of making a siginificant change to the communications, we should change 
that as well.

**Comment by blurfl** [comment]
We don’t know how ugs is bunching its sends. I don’t have a way to packet sniff USB traffic.
I’ll give that GroundControl a try and report back!

**Comment by davidelang** [comment]
manually adding up the L values in this log shows that data is lost as we hit 
>256 characters send (going from 252 to 303), since it then works for a while, 
it looks like the limit test in GC is being applied too late (after the data is 
sent rather than before)

**Comment by BarbourSmith** [comment]
Before I get your hopes up too much, false alarm, I'm still seeing the issue as much as ever.

**Comment by blurfl** [comment]
Yes, I see it too. In fact, the line in error cut off the units digit of the L parameter 🙁

**Comment by blurfl** [comment]
I’d like to try to figure out how to set a flag each time GC sends a line, and wait 10ms? 1ms? 100ms? Before sending the next line. I know, “another flag! 😞”...

**Comment by BarbourSmith** [comment]
This sounds great!

I'm not sure I fully understand.

"data is lost as we hit >256 characters send"

So if the "send" function is called with more than 256 characters? Is that
actually happening?

**Comment by davidelang** [comment]
On Tue, 6 Jun 2017, BarbourSmith wrote:

> "data is lost as we hit >256 characters send"
>
> So if the "send" function is called with more than 256 characters? Is that
> actually happening?

no, I went through the log and totaled up the L entries (96+9+9+...) and we got 
the error when the total went from 252 to 303. So either there was a problem 
with testing the limit (if they had not been processed), or what I think is more 
likely, there was a problem receiving the data as it passed the end of the 
buffer and needed to wrap to the beginning.

We could tell the difference between these if both sender and receiver report 
how much data they think they have in the buffer, and the total number of 
characters sent (we could calculate this if the firmware reports the start and 
end points in the buffer)

**Comment by BarbourSmith** [comment]
Got it.

Good suggestion. Let's look at that.

**Comment by BarbourSmith** [comment]
@blurfl When you have a chance will you try [this Ground Control](https://github.com/MaslowCNC/GroundControl/tree/explore-fast-options-again) and [this firmware](https://github.com/MaslowCNC/Firmware/tree/read-check-sum) and let me know if you have dropped lines when running the magnets file?

That firmware runs the latest changes which fixed some holes in the buffer. That Ground Control runs at full speed again. I don't see any dropped characters in the log after running the magnets test file. 

If that fixes the issue for you too, then we had a lingering bug on the firmware side

@davidelang Good call for suspecting that we were running into something on the firmware side, despite my insisting otherwise

**Comment by blurfl** [comment]
I will do this later this evening.

**Comment by BarbourSmith** [comment]
No rush, I'm about to take off for the day and at the very least it looks like we've got a working version for the release tomorrow 😀 👍 👍

**Comment by blurfl** [comment]
There were lines with lost data. More investigation awaits 😊

**Comment by BarbourSmith** [comment]
Ok, good to know. Thanks for running that test. Unfortunately, because I'm not seeing dropped lines anymore your going to have to be the final word on weather or not we've got the problem solved.

Next I'm going to try to add the .05 second delay proposed in #313 and let's see if that helps

**Comment by blurfl** [comment]
Happy to have the chance to help. A busy day here, but I’ll watch for requests to test :)

**Comment by BarbourSmith** [comment]
Thanks! I'll make sure I've got something I really think might work before asking 👍

**Comment by blurfl** [comment]
I spent some time learning to use wireshark for USB sniffing - works on linux. I was able to capture the USB packets during an instance of this issue. The error sent the machine off in the X direction with no Y coordinate for the gcode.
 Attached are three files:
- a text file of a terminal session during which I ran GC76 and saw the error, search for '**********'
- a pcapng file, the wireshark capture of the USB packets during the session. See the packet numbers called out at '*********' above
- a wireshark text dump of the pcapng file, easier for me to search through than the pcapng file :).

[GCserialIssue.zip](https://github.com/MaslowCNC/GroundControl/files/1079715/GCserialIssue.zip)

**Comment by davidelang** [comment]
The error we are seeing in the communication is that lines are truncated when we 
hit the error, this tends to set the Y coordinate to blank or zero.
k

**Comment by blurfl** [comment]
The packet trace shows that pyserial got the data to the serialUSB driver, and the packet was put onto the bus. Someone more versed in USB than I might be able to dive into the conversation for that transaction.
Looking at the packets, there are several subsequent gcode lines sent before a particular line is acknowledged. I guess that is part of the buffering design, I think a line is echoed on receipt. Perhaps we only need to send one, maybe two lines ahead of the latest line ack'ed.

**Comment by BarbourSmith** [comment]
@blurfl FANTSTIC with the wireshark. Having this insight is really valuable. I'm going to start diging into and trying to understand the files now

**Comment by BarbourSmith** [comment]
My current theory is that we're seeing lost packets when the interrupt for computing the PID loops runs while the serial is reading. 

We saw the frequency of dropped packets decrease significantly when we switched to send new data only once the machine has finished executing the last line, because if the machine is done executing the last line and not doing anything else it will read from the serial port much more quickly resulting in less probability of the interrupt happening during the read.

I'm going to try to test that theory

**Comment by blurfl** [comment]
That sounds promising.

**Comment by BarbourSmith** [comment]
My initial tests seem to say that removing the interrupt call seems to fix the serial issue (unfortunately we need that interrupt to make the motors spin)

It also doesn't seem to be that we need to remove the interrupt call entirely to make the issue go away. Adding back one call to computePID() in the interrupt does not seem to bring the issue back. Calling computePID() for all three axis does. 

It may be that we are simply doing too much from within the interrupt. 

At least we've got a reasonable candidate for where the issue is

**Comment by BarbourSmith** [comment]
@blurfl Based on the information from your packet trace I focused on the the firmware rather than the GC side of things, and I think I've got something worth testing.

I've disabled the timer interrupts for the PID loops while the serial connection is being read. 

When you have a chance, will you test [this firmware](https://github.com/MaslowCNC/Firmware/tree/read-check-sum) with [this Ground Control](https://github.com/MaslowCNC/GroundControl/tree/add-check-sum) and let me know if you see mangled lines?

**Comment by davidelang** [comment]
This indicates that we are not tracking the buffer size correctly (on one side 
or another)

from your traces, can you tell the size of each message? Can you manually go 
through and calculate how much buffer has been used at the time it runs out of 
space?

At this point, I don't know if we know which end is at fault.

There are times when we want to send as much as possible (very small movements 
for example) to keep things smooth, so limiting by the number of lines we send 
ahead is not a great idea.
k

**Comment by BarbourSmith** [comment]
I agree @davidelang I think I've tracked it down to the **hardware** buffer on the Arduino side. I had ruled that out because in one pull request someone increased it to size 256, but it turns out you can't increase it from with the sketch, you have to edit the files with the arduino ide installation directory as described [here](https://internetofhomethings.com/homethings/?p=927) in the section entitled "Hardware Serial Buffer Expansion". By doing that I can make the problem go away (I would **love** to see someone else verify this)

Asking everyone to edit their Arduino install isn't a viable option, so we're going to have to focus on either emptying the buffer more often on the firmware side, or dumping less data into it suddenly on the Ground Control side.

**Comment by blurfl** [comment]
I'm sorry to say that we're still losing characters.
 Your addition of a character count and check are very handy, though. I can tail log.txt to a grep for 'Chars lost' and catch the ones that aren't big enough to send the sled way off base.
 Below are the lines caught in running "grep -A 1 'Chars lost' " on the first 66 lines of a gcode file. I quit after that and tried some other files, they showed the same.


1-->>Chars lost in serial -->>
G00 Z0  W12 

19-->>Chars lost in serial -->>
G1 X0.02103 Y0.006

65522-->>Chars lost in serial -->>
0.01041 F20.0  W34 

60-->>Chars lost in serial -->>
G1 X0.09543 Y0.07788 F20.0  W3

28-->>Chars lost in serial -->>
G1 X0.00638 Y-0.02103 F20.0 

33-->>Chars lost in serial -->>
G1 X0.11792 Y-0.03561 F20.0  W34 

Note that short lines like 'G00 Z0  W12' seem to always trigger the message, line length mis-counted?

I've attached the test~~~.nc file I was using, the log~~~.txt file made by GC and the Serial~~~.txt captured on the serial Rx pin of the Arduino (AUX5, very handy).

[testing-GroundControl-add-check-sum.zip](https://github.com/MaslowCNC/GroundControl/files/1082347/testing-GroundControl-add-check-sum.zip)

**Comment by blurfl** [comment]
I think the buffer sizes can be changed by the script, as is currently done in Firmware. The HardwareSerial.h file does mention some issues, though. I've copied part of the file here, ending with one of the if statements that looks to me like it would permit changing the buffer value.


// Define constants and variables for buffering incoming serial data.  We're
// using a ring buffer (I think), in which head is the index of the location
// to which to write the next incoming character and tail is the index of the
// location from which to read.
// NOTE: a "power of 2" buffer size is reccomended to dramatically
//       optimize all the modulo operations for ring buffers.
// WARNING: When buffer sizes are increased to > 256, the buffer index
// variables are automatically increased in size, but the extra
// atomicity guards needed for that are not implemented. This will
// often work, but occasionally a race condition can occur that makes
// Serial behave erratically. See https://github.com/arduino/Arduino/issues/2405
#if !defined(SERIAL_TX_BUFFER_SIZE)

**Comment by davidelang** [comment]
a couple things here.

1. setting the buffer size to 512 didn't solve the problem, so with the issues 
that Scott is identifying, let's go ahead and set it back to 256 for now.

2. with no flow control, interrupts that take place in the middle of receiving 
data will cause problems, because the sender has no way of knowing to pause.

so I think we need to look at the interrupt routines and see if we can slim them 
down, we want to do as little as possible during the interrupt, doing more 
complex things in the main loop.

**Comment by blurfl** [comment]
The default HardwareSerial Tx and Rx buffer size is 64

**Comment by blurfl** [comment]
The [Arduino serial issue](https://github.com/arduino/Arduino/issues/2405) mentioned above discusses atomicity and interrupts when updating buffer head and tail pointers. Could we be getting bitten by encoder interrupts while updating _our_ buffer?

**Comment by BarbourSmith** [comment]
@blurfl  is right that the default Rx buffer size is 64 and @davidelang is right that our internal buffer size is is 512. The difference between the two is what's causing the issue. If we don't empty the Arduno's 64 byte buffer into our 512 byte buffer fast enough, Ground Control keeps sending data (because there is space in the 512 byte buffer) and overflows the 64 byte buffer. 

I've added a print statement to the latest versions of those two branches linked above which lets us know when the buffer has overflowed (if it's the default size).

@blurfl When you were still seeing characters lost above, was that after editing the file HardwareSerial.h in the Arduino installation folder [https://internetofhomethings.com/homethings/?p=927](https://internetofhomethings.com/homethings/?p=927) and then restarting the Arduino IDE? If so we may have multiple issues going on

**Comment by blurfl** [comment]
@BarbourSmith the file HardwareSerial.h in my installation is still the default 64. 
 Note, though, that adding a print statement to the sketch to see the values of SERIAL_TX_BUFFER_SIZE and SERIAL_RX_BUFFER_SIZE reveals that they are altered by the #define statements in the sketch - the values are 64 when the #defines are commented out, 512 when the #defines are active. 
 Are there additional tests you would like to see? maybe:
- comment out the #define lines in the sketch but do not edit HardwareSerial.h (leave at 64)
- edit HardwareSerial.h to 256 and comment out the #define lines in the sketch
- edit HardwareSerial.h to 512 and comment out the #define lines in the sketch
Let me know what would be most useful, I'll be glad to run the needed tests.

**Comment by BarbourSmith** [comment]
Yes! Those tests sound like exactly what we need to know! 👍 👍 👍 

I think the 

- edit HardwareSerial.h to 512 and comment out the #define lines in the sketch

is probably the most urgent test to run, but each of the three will give us some new an important information.

Great suggestions!

**Comment by blurfl** [comment]
Roger that, will test 'edit HardwareSerial.h to 512 and comment out the #define lines in the sketch' first and report back.

**Comment by davidelang** [comment]
I believe that if we do the #define after loading the library (it may be needed 
to do a #undefine first) that our value will be used by the library.

it looks like you do need to #undefine, see 
https://stackoverflow.com/questions/42151118/how-to-override-define-in-arduino-library-header

**Comment by blurfl** [comment]
The test 'edit HardwareSerial.h to 512 and comment out the #define lines in the sketch' still results in 'Chars lost...'.
I'll try 256 next, yes?

**Comment by BarbourSmith** [comment]
Hmm 😞 

I think trying 256 is a good thing to try next. 

Did you see any difference in the number of times that chars were lost or in the way that they appeared to be lost?

**Comment by blurfl** [comment]
The test 'edit HardwareSerial.h to 256 and comment out the #define lines in the sketch' still results in 'Chars lost...'.
Also saw messages about 'Buffer overflow'.

Note that both test ...512 and ...256 had far fewer instances of 'Chars lost'.

**Comment by BarbourSmith** [comment]
I'm not sure we can 100% trust the "Chars lost" print statement to tell us what's really going on. It's a good first pass, but I think it's worth double checking the lines it catches to make sure that we're really loosing data. 

From you original post I would say 

**19-->>Chars lost in serial -->>
G1 X0.02103 Y0.006**

is a pretty clear example of data really being lost, while 

**33-->>Chars lost in serial -->>
G1 X0.11792 Y-0.03561 F20.0 W34**

Doesn't make much sense. The print statement is telling us that we lost another 33 characters, but I'm not sure what those characters would have been because the 'W34' is added to the very end of the string on the python side so I don't think we've actually lost characters

**Comment by blurfl** [comment]
I see what you mean. The previous and subsequent lines look correct, and these lines look correct in the serial log generated at the point of Serial.read(c ) as well. The 'chars lost' message is generated after the characters are retreived from Serial.read(c).

**Comment by BarbourSmith** [comment]
Weirdly enough, I think what's tripping up the 'chars lost' calculation is that the built in Arduino string length function is sometimes returning the wrong length. Why that would happen is a mystery to me.

When you have the HardwareSerial.h set to 256 are you still seeing lines that are truncated in the way where it's clear we've really lost data?

**Comment by blurfl** [comment]
I wasn’t capturing and I stopped after the first instance and repeated the run  I’ll do a full capture at 256 in HardwareSerial.h and report back.

**Comment by blurfl** [comment]
Set to 256 I again saw one ‘Chars lost’ message, but every line echoed back to GC looked full (ended with a ‘Wxx’) including the one that triggered the message. Go figure…
Interesting aside, I had altered the #define for VERSIONNUMBER to a string with the branch name in it, and that change caused the Mega to mis-report the sled position when GC starts. Changing to a shorter string works correctly.

**Comment by BarbourSmith** [comment]
Great! So it seems like the work around is to change the arduino buffer size, now we've got to find a way to get the same behavior without editing files outside our system.

That is a very interesting aside. I wonder what caused that.

**Comment by blurfl** [comment]
The aside was caused by the extra length of the #define VERSIONNUMBER string stomping on the beginning of EEPROM storage. Moving the xx_EEPROM_ADR values up by 20 each restored normal startup. We need to be careful about the length of the VERSIONNUMBER string...😉

**Comment by GeroBH** [comment]
If you have more than one arduino project like me, this might be of interest [http://www.hobbytronics.co.uk/arduino-serial-buffer-size](url)
You can make 'Boards' for several buffer sizes for testing without altering the original files.

**Comment by davidelang** [comment]
could someone post a full log showing the failures?

does this log include the number of characters in the buffer on the firmware 
side?

I'd like to try and rule out the possibility that we are sending more characters 
than there is room for in the buffer before we worry too much about expanding 
the buffer size and/or doing drastic surgery on other things.

**Comment by davidelang** [comment]
it could be loosing the newline.

I pointed out an issue before where the encoding of the length of the message 
could cause it to be a character longer than it calculates, did that ever get 
fixed?
k

**Comment by blurfl** [comment]
I've attached logs from a run that had two lines incomplete, missing the Wxx value at the end. The setup for this run was a buffer size of 256 set by editing HardwareSerial.h in the core.

The files are:
1. GCconsole-256inCore.txt is the complete console output of the  GC session
2. log-256inCore.txt is the log written to disk by GC
3. foundG0G1-log-256inCore.txt is the result of “grep ‘G0\|G1’ log-256inCore.txt”, ignoring the non-gcode lines

[256inCore.zip](https://github.com/MaslowCNC/GroundControl/files/1082852/256inCore.zip)

**Comment by BarbourSmith** [comment]
@davidelang I don't think that did get fixed. What exactly were the conditions under which that would happen?


@blurfl 
So does that mean that we are sure we're still seeing issues even after changing the buffer size to 256 in HardwareSerial.h?

I think the corrupted lines there might have come from the internal buffer overflowing where we see lines like this:

```
buffer overflow
Buffer begin: 42
Buffer end: 41
```

Rather than from the arduino Hardware buffer overflowing

One reason that we might be seeing that is that the line lengths being sent don't include the extra charicters we're adding to record the line length ie the **W34** on the end of the line takes up space in the buffer which we're not accounting for. I've just been leaving mine at 512 for now so that we don't have to worry about it until we know the hardware buffer is OK

**Comment by davidelang** [comment]
I posted this in one of the threads, but am not seeing it now.

In GC in serialPortThread.py, the line:

        message = message + ''L' + str(len(message) + 1 + 2 + len(str(len(message))) ) + " \n"

can be off by one if the +1 +2 causes len(str(len(message))) to be different than (len(str(len(message?)+1+2)) (say len(message) = 7 , then you output L10, but the actual length is 11

I'm assuming that in your testing branch you have the L## enabled for debugging (I haven't had a chance to dig into the log dump posted earlier, and probably won't until the morning)

**Comment by blurfl** [comment]
@BarbourSmith will the logs that I uploaded answer whether the hardware buffer overflowed? I don’t think they will, I’d have to capture the characters as they are grabbed from there by Serial.read(c). I was printing those to Serial2 a couple days ago, could rig to do that again as well if needed. 
I would observe that when the hardware buffer is at the default 64, I see more incomplete lines. Changing the size of the hardware buffer shouldn't affect an issue with the software buffer, yes? 
What processing could be taken out of the interrupt service routines?

**Comment by davidelang** [comment]
since the GC dsoftware is paceing it's sending based on a 256 byte buffer, a 
hardware buffer of 64 will be much easier to overflow.

we should be paceing to avoid overflow of the hardware buffer, because we don't 
know how fast it can be drained.

the easiest thing to do is to set the hardware buffer to be the same size as the 
software buffer and not assume that anything that hasn't been acked has been 
read from the hardware buffer (in fact, if the code could work strictly on the 
hardware buffer and not maintain it's own input buffer, there would be 
advantages, but we would have the problem of needing to 'peek ahead' to watch 
for a stop message.

**Comment by blurfl** [comment]
Here are four runs  ~-64 with buffersize 64 and the rest at 256. The .nc file is included as well.
In the HardwareSerial files, the incoming characters were echoed out Serial3 as they were read in. Oddly enough, some lines of those files are missing the initial 'Gx ' portion although the corresponding lines in the log written by GC as received from the firmware shows those lines intact. The four runs  ~-256 show these quirks at different places; I ran multiple times using different baud rates for Serial3 and different methods to capture the stream. Thinking that there might be a conflict between the hardware serial buffer and the EEPROM space, runs 3 and 4 had the EEPROM starting addresses raised.

[0618-64.zip](https://github.com/MaslowCNC/GroundControl/files/1083374/0618-64.zip)
[0618-256-run1.zip](https://github.com/MaslowCNC/GroundControl/files/1083375/0618-256-run1.zip)
[0618-256-run2.zip](https://github.com/MaslowCNC/GroundControl/files/1083376/0618-256-run2.zip)
[0618-256-run4.zip](https://github.com/MaslowCNC/GroundControl/files/1083377/0618-256-run4.zip)
[Magnet array 12mm x 6cm x 5-SAE.nc.zip](https://github.com/MaslowCNC/GroundControl/files/1083392/Magnet.array.12mm.x.6cm.x.5-SAE.nc.zip)

**Comment by BarbourSmith** [comment]
These issues have been fixed (thank goodness!)
